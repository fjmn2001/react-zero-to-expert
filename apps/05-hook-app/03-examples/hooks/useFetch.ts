import { useEffect, useState } from "react"

const useFetch = (url: string) => {
  const [state, setState] = useState({ data: null, loading: true, error: null })

  useEffect(() => {
    setState({ loading: false, error: null, data: null })

    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        setState({
          loading: false,
          error: null,
          data,
        })
      })
  }, [url])

  return state
}

export default useFetch
