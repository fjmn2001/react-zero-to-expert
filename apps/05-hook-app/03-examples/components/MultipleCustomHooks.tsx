import "./examples.css"
import useFetch from "../hooks/useFetch"
import { useCounter } from "../../01-useState/hooks/useCounter"

const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1)
  const { loading, data } = useFetch(
    `https://breakingbadapi.com/api/quotes/${counter}`
  )
  const { author, quote } = (!!data && data[0]) as unknown as {
    author: string | undefined
    quote: string | undefined
  }

  return (
    <>
      <h1>Breaking Bad Quotes</h1>
      <hr />

      {loading ? (
        <div className={"alert alert-info text-center"}>Loading...</div>
      ) : (
        <blockquote className={"blockquote text-right"}>
          <p className={"blockquote-body"}>{quote}</p>
          <footer className={"blockquote-footer"}>{author}</footer>
        </blockquote>
      )}

      <button className={"btn btn-primary"} onClick={() => increment()}>
        Next Quote
      </button>
    </>
  )
}

export default MultipleCustomHooks
