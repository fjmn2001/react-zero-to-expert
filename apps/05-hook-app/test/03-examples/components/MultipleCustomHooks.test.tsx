import { describe, expect, MockedFunction, test, vi, beforeEach } from "vitest"
import { render } from "@testing-library/react"
import MultipleCustomHooks from "../../../03-examples/components/MultipleCustomHooks"
import useFetch from "../../../03-examples/hooks/useFetch"

vi.mock("../../../03-examples/hooks/useFetch")

describe("Testing MultipleCustomHooks.tsx", () => {
  beforeEach(() => {
    vi.resetAllMocks()
  })

  test("it should show correctly", () => {
    ;(useFetch as MockedFunction<any>).mockReturnValue({
      data: null,
      loading: true,
      error: null,
    })
    const { container } = render(<MultipleCustomHooks />)
    expect(container).toMatchSnapshot()
  })

  test("it should show information", () => {
    ;(useFetch as MockedFunction<any>).mockReturnValue({
      data: [{ author: "Francisco", quote: "typescript" }],
      loading: false,
      error: null,
    })
    const { container } = render(<MultipleCustomHooks />)
    expect(container).toMatchSnapshot()
    expect(container.getElementsByClassName("blockquote").length).toBe(1)
  })
})
