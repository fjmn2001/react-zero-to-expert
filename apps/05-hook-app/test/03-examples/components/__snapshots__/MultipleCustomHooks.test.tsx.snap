// Vitest Snapshot v1

exports[`Testing MultipleCustomHooks.tsx > it should show correctly 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    class="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    class="btn btn-primary"
  >
    Next Quote
  </button>
</div>
`;

exports[`Testing MultipleCustomHooks.tsx > it should show information 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <blockquote
    class="blockquote text-right"
  >
    <p
      class="blockquote-body"
    >
      typescript
    </p>
    <footer
      class="blockquote-footer"
    >
      Francisco
    </footer>
  </blockquote>
  <button
    class="btn btn-primary"
  >
    Next Quote
  </button>
</div>
`;
